SC2: Rework Mercenary items

Snarkie opened the pull request:
## What is this fixing or adding?

This is a rework to Mercenary items, the main focus is on splitting the Mercenary items between Terran and Zerg races.

> Quote: changed/removed in later revision

### Changes to existing items:
- **Rogue Forces**, Mercenary calldowns are no longer limited by charges:
  - Only affects Terran Mercenaries now.
- **Rapid Reinforcement**, Reduces cooldowns of mercenary calldowns by 60s:
  - Only affects Terran Mercenaries now.

>-  **Mercenary Munitions**,  Increases attack speed of all combat units by 15%:
>   - Only affects Terran Mercenaries now (instead of all Terran units).
>   - Nerfed to 10% attack speed.

- **Fast Delivery**, Mercenary calldowns can be deployed right at the mission start:
  - Is now progressive.
  - Only affects Terran Mercenaries
  - Level 1: One Mercenary can be deployed right at the mission start.
 > - Level 2: All Mercenary calldowns can be deployed right at the mission start.
- Level 2: 3 Mercenary calldowns can be deployed right at the mission start.

### New items:
- **Unrestricted Mutation:** Zerg Mercenaries are no longer limited by charges.
- **Cell Division:** Adds additional units to Zerg Mercenaries.
  - does not affect Torrasque or Yggdrasil.
  - adds 2 Devouring Ones, 1 additional merc for all other Zerg Mercenaries.
  - does not increase cost (besides supply).
- **Evolutionary Leap:** 
> - Progressive.
  - Level 1: Halves initial cooldowns for Zerg Mercenaries.
>  - Level 2: Removes initial cooldowns for Zerg Mercenaries.
- **Self Sufficient:** Removes the supply cost for Zerg Mercenaries.
  - if that proves unhealthy, can be changed to halve the supply cost.
 
### Other Changes:

- Small balance pass on Mercenaries:
  - You can only control 1 Torrasque and Jackson's Revenge each, even with infinite charges.
  - Torrasque armor 5 -> 3, Cooldown 600s -> 420s
  - Midnight Raider (Liberator) Cooldown 300s -> 360s
- Streamlined Mercenary tooltips
![grafik](https://github.com/user-attachments/assets/92353d48-c6a7-4a9a-825c-7e229abaa242)
- Removed hidden 2nd ability for Devouring Ones (Zerglings). This was used for Swarmling upgrade, but was causing multiple bugs and display issues, for example incorrect counter display for the predator nest, or incorrect cooldown manipulations on the zerglings (since it is a global player cooldown, any changes would apply twice).

## How was this tested?

Local test map, local campaigns, tested some specific missions for their interactions with mercs and their cooldowns

### Notes
> - renames the existing Terran Merc items.
- requires some refactoring of the Merc trigger system.
[
Data PR](https://github.com/Ziktofel/Archipelago-SC2-data/pull/367)

Snarkie committed:
Rework Mercenary items

MatthewMarinets commented:
# Compatibility
For the compatibility with older clients, you'd want to give Unrestricted Mutation for free if the API level is < 4 (ie generated on current live) and the player has Rogue Forces. To do that:
* Update `calculate_items()` in client.py
* under the `if ctx.slot_data_version < 4:` if statement (around line 1265). add something like:
```python
        if item_data[item_names.ROGUE_FORCES].code in set(item.item for item in ctx.items_received if item.player == ctx.slot):
            items.append(NetworkItem(item_data[item_names.UNRESTRICTED_MUTATION].code, 0, 0, 0))
```
(this will require defining `item_list` earlier in the function;  the definition around line 1282 can be pulled up to the top of the function)

# Balance
I like the direction, though still worry a bit.
* Mercenary Munitions can probably be hit a little harder by the nerf, I'd say 5% is a good place to be given how many mercs there are that can be affected and how strong they already are.
* I'm unhappy with the cost reduction on Cell Division; that's a powerful upgrade that is also doubling as a powerful resource efficiency upgrade. I think it is pre-emptively trying to buff an item that doesn't need a buff, and that it's probably better to let people get it and play with it and provide feedback on if the increased cost per round of mercs is strongly felt (I don't think it will be).
* I'm worried about level 2 of fast delivery as well. It's level 2 of a progressive, and will thus be more uncommon, but it also means players are most likely to get it when they have many mercs and thus the item will be at its most OP. I'd instead advocate for having level 2 just double the power of level 1 -- give 2 instant merc charges at the start, and the rest normal.
* Similar thoughts for the zerg equivalent, though a bit balanced out as they only have 3 mercs I'd call very strong (hunter killers, thornshells, torrasque). I'm willing to say we should allow it in and let people play with it, but when I'm feeling conservative I ask why we even need level 2 at all

Ziktofel commented:
```
Mercenary Munitions, Increases attack speed of all combat units by 15%:

    Only affects Terran Mercenaries now (instead of all Terran units).
    Nerfed to 10% attack speed.

```
I think I'd rather find an another name as this clashes with Coop Raynor (borrowing a name for an another effect that already exists in game)

5% is Ultra-Capacitors (and stacks per weapon upgrade level) - Exists in vanilla WoL

```
I'm unhappy with the cost reduction on Cell Division; that's a powerful upgrade that is also doubling as a powerful resource efficiency upgrade. I think it is pre-emptively trying to buff an item that doesn't need a buff, and that it's probably better to let people get it and play with it and provide feedback on if the increased cost per round of mercs is strongly felt (I don't think it will be).
```
The thing is with potential trappiness of the item (a beneficial item shouldn't have trap-like behavior like hampering your attack timing)

```
Fast Delivery, Mercenary calldowns can be deployed right at the mission start:

    Is now progressive.
    Level 1: One Mercenary can be deployed right at the mission start.
    Level 2: All Mercenary calldowns can be deployed right at the mission start.

```
How does that work? If none mercs aren't on cooldown at the start but requesting first merc puts everything on cooldown, then it's a potential trap (delays all other mercs by the time you purchase the first one)

Snarkie commented:
> How does that work? If none mercs aren't on cooldown at the start but requesting first merc puts everything on cooldown, then it's a potential trap (delays all other mercs by the time you purchase the first one)

I did make sure the cooldowns get reset to the value they would have had if you never used a merc up to that point. Went through several iterations for it to work properly, currently the cooldown is tracked with a hidden long cooldown dummy ability, which is used to account for triggers freezing global cooldowns often used in campaign missions.

Snarkie commented:
> I think I'd rather find an another name as this clashes with Coop Raynor (borrowing a name for an another effect that already exists in game)

I don't like this, the effect was changed to affect mercs specifically because of the name (and because it needed a nerf). If we don't keep the name, I don't see a point in having the upgrade for mercs at all. They don't need an attackspeed increase, and it is bland and boring.

So in my opinion, either we keep Mercenary Munitions affecting mercenaries, or we nerf/remove the original item in a different way and make something completely different that's more interesting for the mercs.

Snarkie committed:
fix compatibility for Rogue Forces

Snarkie committed:
fix typo

MatthewMarinets commented:
> I think I'd rather find an another name as this clashes with Coop Raynor (borrowing a name for an another effect that already exists in game)

I agree with Snarky on this one, we have no incentive to ever merge the co-op Raynor upgrade as it's just plain uninteresting. The name itself is what caused us to go in the direction, so I think it's better to keep the name for an upgrade that affects mercenaries.

> > I'm unhappy with the cost reduction on Cell Division; that's a powerful upgrade that is also doubling as a powerful resource efficiency upgrade. I think it is pre-emptively trying to buff an item that doesn't need a buff, and that it's probably better to let people get it and play with it and provide feedback on if the increased cost per round of mercs is strongly felt (I don't think it will be).
> 
> The thing is with potential trappiness of the item (a beneficial item shouldn't have trap-like behavior like hampering your attack timing)

Noting that there was some discussion in the discord today about making some of these global upgrades require a research to become effective. Could we make the extra merc in the calldown only apply once the player has opted in in-game, by doing a research on the predator nest or evo chamber? There was discussion of having an option when we do this generally; in this case as we always want an opt-in, the option could control between a free-and-instant opt-in or toggle vs having to spend resources and research time to get it.

As we don't have such an option yet, I propose that we remove the free resource efficiency from Cell Division, and to avoid nerfing timings we only give the player the extra charges only after they research a free 5s upgrade at the evo chamber.

Snarkie committed:
Rework Mercenary items revision 1

Snarkie commented:
## Revision 1
- Reverted Mercenary Munitions, this PR no longer changes it (besides some re-ordering)
- Added a new item, Signal Beacon (Terran): Makes Terran mercenaries deploy at the rally point instantly
- Removed Evolutionary Leap level 2, it is no longer progressive (still progressive in code, but not visible to end user)
- Changed Fast Delivery level 2 to allow up to 3 instant calldowns, instead of all
- Added an icon indicator for Fast Delivery
![GIF_28 02 2025_15-10-32](https://github.com/user-attachments/assets/21a63d82-eee8-4f11-b1c1-dff4d0db6c34)

- Removed cost reduction for Cell Division, additional units increase the costs now
- Added a toggleable ability to deactivate the effect of Cell Division, restoring original unit count and costs
- Balance changes to Zerg mercenaries (see next chapter)
- Renamed Merc Ultralisk to Wise Old Torrasque
- Changed the group border color for most of the Zerg mercs to white (like in Broodwar)

## Balance Changes:

- ### Torrasque:
  - Cost 500/300 -> 600/400
  - Armor 5 -> 3
  - Revive cooldown 60s -> 180s baseline, reverts to 60s with Torrasque Strain item
- ### Hunter Killer 
  - Cost 250/100 -> 350/150
- ### Thornshell 
  - Cost 150/75 -> 250/125
- ### Devouring Ones 
  - Cost 200 -> 300

Snarkie committed:
Fix evolutionary leap progressive
